src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -1704,8 +1707,24 @@ MarkRanges MergeTreeDataSelectExecutor::filterMarksUsingIndex(
 
                 if (index_helper->isVectorSimilarityIndex())
                 {
-                    auto rows = condition->calculateApproximateNearestNeighbors(granule);
-
+                    read_hints.ann_search_results = condition->calculateApproximateNearestNeighbors(granule);
+
+                    /// corresponding ranges have to be returned in ascending order
+                    auto rows = read_hints.ann_search_results.value().first;
+                    std::sort(rows.begin(), rows.end());
+#ifndef NDEBUG
+                    /// Duplicates should in theory not be possible but who knows ...
+                    const bool has_duplicates = std::adjacent_find(rows.begin(), rows.end()) != rows.end();
+                    if (has_duplicates)
+                        throw Exception(ErrorCodes::INCORRECT_DATA, "Usearch returned duplicate row numbers");
+#endif
+
+                    if (settings[Setting::vector_search_with_rescoring] ||
```

Comment:
I don't understand why we reset `read_hints` in l. 1726. Will bad things happen if we don't reset?